app/services/import_polygons.rb

```diff
@@ -41,4 +43,57 @@ def location_categories_include?(region_name)
       location_type == :london_boroughs && DOWNCASE_BOROUGHS.include?(region_name) ||
       location_type == :cities && DOWNCASE_CITIES.include?(region_name)
   end
+
+  def get_buffers(points)
+    # ArcGIS API documentation: https://ons-inspire.esriuk.com/arcgis/sdk/rest/index.html#//02ss0000003z000000
+    buffer_distances_in_miles = [5, 10, 15, 20, 25]
+    buffered_boundaries = {}
+    buffer_distances_in_miles.each do |distance|
+      # convert 1D array, to 2D array for arcgis
+      polygon_coords = points.each_slice(2).to_a
+      response = HTTParty.get(buffer_api_endpoint(polygon_coords, convert_miles_to_metres(distance)))
+      buffer_coords = response.dig("geometries", 0, "rings").flatten
+      # algolia requires a 2D containing 1 element with all the coordinates contained within it
+      # so buffer_coords array has been wrapped in another array
+      buffered_boundaries[distance.to_s] = format_buffer_coords_for_algolia(buffer_coords)
+    end
+    buffered_boundaries
+  end
+
+  def buffer_api_endpoint(coords, distance)
+    geometries_param = {
+      'geometryType' => 'esriGeometryPolygon',
+      'geometries' => [{ 'rings' => [coords] }]
+    }
+
+    params = { 'geometries' => geometries_param.to_s,
+               'inSR' => '4326',
+               'outSR' => '4326',
+               'bufferSR' => '3857',
+               'distances' => distance.to_s,
+               'unit' => '',
+               'unionResults' => 'true',
+               'geodesic' => 'false',
+               'f' => 'json' }.to_param
+
+    buffer_api_url = 'https://ons-inspire.esriuk.com/arcgis/rest/services/Utilities/Geometry/GeometryServer/buffer?'
+
+    api_endpoint = buffer_api_url + params
+
+    if api_endpoint.length > 2000
+      # Reduce number of coordinates in order to have fewer than 2000 characters in request url
+      coords = coords.each_with_index
+      .map { |item, index| item if (index % 5).zero? }
```

Comment:
I don't feel particularly comfortable downsampling the initial polygon this crudely unless we can clearly show it is acceptable for all cases. Not sure how everyone else feels

Reply:
@david-mears-dfe

Reply:
We tested this by checking two buffered polygons which have high numbers of points in their original boundaries (1000s).

```ruby
# These numbers should be halved in order to get the number of points; these are the numbers of lats and lngs.
pry(main)> LocationPolygon.with_name('portsmouth').boundary.length
=> 1886
[2] pry(main)> LocationPolygon.with_name('wakefield').boundary.length
=> 4410
```

<img width="1193" alt="Screenshot 2020-11-25 at 11 28 35" src="https://user-images.githubusercontent.com/60350599/100865132-dba5ab00-348e-11eb-99d8-ed6d2c66801e.png">

<img width="1219" alt="Screenshot 2020-11-25 at 11 32 20" src="https://user-images.githubusercontent.com/60350599/100865135-dd6f6e80-348e-11eb-829b-b520e2e19b70.png">

I think that going down from 1000s to 100s of points (or even less) would not result in an appreciable difference in the shape of the buffer.

@josephhull676 Would you be able to tell us what is the length of the buffers displayed above (I forget if they were 1 miles or 5 miles), and what is the shortest length of any buffer? Something like:

```ruby
LocationPolygon.pluck(:buffers) # will give you only the buffers column, as an array.
  .map { |polygon_buffers_list|
    polygon_buffers_list.map(&:length) # or is it :&? will give you a list of the length of each buffer on this polygon
  }.flatten # flatten all the buffer lengths into a 1-D array
  .sort # get them in length order
```

Reply:
I **think** the Portsmouth buffer was 5 miles and the Wakefield buffer was 10 miles.

The 5 mile buffer polygon for Portsmouth is made of 238 points.

The 10 mile buffer polygon for Wakefield is made of 264 points.

The smallest amount of points for any buffer is 216.

Reply:
Given we're downsampling the polygon for the purpose of creating buffers (and the accuracy of these is far less important) and not downsampling the base polygon itself, this is ok